src/ILLink.Shared/TrimAnalysis/StackLattice.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using ILLink.Shared.DataFlow;
+
+namespace ILLink.Shared.TrimAnalysis
+{
+	// A lattice over stacks where the stored values are also from a lattice.
+	public readonly struct StackLattice<TValue, TValueLattice> : ILattice<StructStack<TValue>>
+		where TValue : IEquatable<TValue>
+		where TValueLattice : ILatticeWithUnknownValue<TValue>
+	{
+		public readonly TValueLattice ValueLattice;
+
+		public StructStack<TValue> Top { get; }
+
+		public StackLattice (TValueLattice valueLattice)
+		{
+			ValueLattice = valueLattice;
+			Top = new StructStack<TValue> ();
+		}
+
+		public StructStack<TValue> Meet (StructStack<TValue> left, StructStack<TValue> right)
+		{
+			// Meet(value, Top) = value
+			if (left.Equals (Top)) return new StructStack<TValue> (new StructStack<TValue> (right));
+
+			if (right.Equals (Top)) return new StructStack<TValue> (new StructStack<TValue> (left));
+
+			if (left.Count != right.Count) {
+				// Force stacks to be of equal size to avoid crashes.
```

Comment:
We discussed this with Jeremi, but I wasn't sure it's actually doable. The problem is that it happens during Meet. You have two blocks which produce output state which has different stack depth. They both are predecessors to some other block. We run Meet on the states of all predecessors, and the depth discrepancy will be visible inside the Meet. Without this, we can't Meet those stacks. And scanner will not know about this since it sees each block on its own and it won't know that the stack depth is different from what it should be.

Reply:
Just a note that it is invalid IL if stack heights are different. III.1.7.5 in the ECMA spec prescribes the stack height and types of elements on the stack need to be the same on control merge points. We could also abort analyzing the method by throwing a catchable exception if that's more convenient. Invalid IL is an exceptional state.

Reply:
We were trying to mimic the existing behavior (which will push unknowns in merge points). For correctness we would have to generate a warning if we give up on analysis though. For obfuscated assemblies, it might be relatively common to have unbalanced stacks and thus we could produce quite a few warnings for them, which we didn't before.
I'm not opposed to doing it though - we don't support processing obfuscated assemblies anyway.

Reply:
Removed LatticeWithUnknownValue. Added a warning indicating invalid IL.